File: bindings/dotnet/FireLocal/FieldValue.cs
using System;
using System.Collections.Generic;

namespace FireLocal
{
    /// <summary>
    /// FieldValue helpers for special Firestore-compatible operations
    /// </summary>
    public static class FieldValue
    {
        /// <summary>
        /// Get current server timestamp
        /// </summary>
        /// <returns>FieldValue for server timestamp</returns>
        public static Dictionary<string, object> ServerTimestamp()
        {
            return new Dictionary<string, object>
            {
                ["_firelocal_field_value"] = "serverTimestamp",
                ["value"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
            };
        }

        /// <summary>
        /// Increment a numeric field
        /// </summary>
        /// <param name="n">Amount to increment by</param>
        /// <returns>FieldValue for increment operation</returns>
        public static Dictionary<string, object> Increment(long n)
        {
            return new Dictionary<string, object>
            {
                ["_firelocal_field_value"] = "increment",
                ["value"] = n
            };
        }

        /// <summary>
        /// Add elements to an array (unique)
        /// </summary>
        /// <param name="elements">Elements to add</param>
        /// <returns>FieldValue for arrayUnion operation</returns>
        public static Dictionary<string, object> ArrayUnion(params object[] elements)
        {
            return new Dictionary<string, object>
            {
                ["_firelocal_field_value"] = "arrayUnion",
                ["value"] = elements
            };
        }

        /// <summary>
        /// Remove elements from an array
        /// </summary>
        /// <param name="elements">Elements to remove</param>
        /// <returns>FieldValue for arrayRemove operation</returns>
        public static Dictionary<string, object> ArrayRemove(params object[] elements)
        {
            return new Dictionary<string, object>
            {
                ["_firelocal_field_value"] = "arrayRemove",
                ["value"] = elements
            };
        }

        /// <summary>
        /// Delete a field from a document
        /// </summary>
        /// <returns>FieldValue for delete operation</returns>
        public static Dictionary<string, object> Delete()
        {
            return new Dictionary<string, object>
            {
                ["_firelocal_field_value"] = "delete"
            };
        }
    }
}

File: bindings/python/firelocal/field_value.py
"""
FieldValue helpers for special operations
"""

import json
import time
from typing import Any, List


def server_timestamp() -> dict:
    """
    Get current server timestamp
    
    Returns:
        FieldValue for server timestamp
    """
    return {
        "_firelocal_field_value": "serverTimestamp",
        "value": int(time.time() * 1000),  # milliseconds
    }


def increment(n: int) -> dict:
    """
    Increment a numeric field
    
    Args:
        n: Amount to increment by
        
    Returns:
        FieldValue for increment operation
    """
    return {
        "_firelocal_field_value": "increment",
        "value": n,
    }


def array_union(elements: List[Any]) -> dict:
    """
    Add elements to an array (unique)
    
    Args:
        elements: Elements to add
        
    Returns:
        FieldValue for arrayUnion operation
    """
    return {
        "_firelocal_field_value": "arrayUnion",
        "value": elements,
    }


def array_remove(elements: List[Any]) -> dict:
    """
    Remove elements from an array
    
    Args:
        elements: Elements to remove
        
    Returns:
        FieldValue for arrayRemove operation
    """
    return {
        "_firelocal_field_value": "arrayRemove",
        "value": elements,
    }


def delete_field() -> dict:
    """
    Delete a field from a document
    
    Returns:
        FieldValue for delete operation
    """
    return {
        "_firelocal_field_value": "delete",
    }
